python/cudf/cudf/core/indexing_utils.py

```diff
@@ -241,3 +267,356 @@ def parse_row_iloc_indexer(key: Any, n: int) -> IndexingSpec:
+
+    This converts label-based lookups into structured positional
+    lookups.
+
+    Valid values for the key are
+    - a slice (endpoints are looked up)
+    - a scalar label
+    - a boolean mask of the same length as the index
+    - a column of labels to look up (may be empty)
+
+    Parameters
+    ----------
+    key
+        Key for label-based row indexing
+    index
+        Index to act as haystack for labels
+
+    Returns
+    -------
+    IndexingSpec
+        Structured information for indexing
+
+    Raises
+    ------
+    KeyError
+        If any label is not found
+    ValueError
+        If labels cannot be coerced to index dtype
+    """
+    n = len(index)
+    if isinstance(key, slice):
+        return find_label_range_or_mask(key, index)
+    else:
+        is_scalar = _is_scalar_or_zero_d_array(key)
+        if is_scalar and isinstance(key, np.ndarray):
+            key = cudf.core.column.as_column(key.item(), dtype=key.dtype)
+        else:
+            key = cudf.core.column.as_column(key)
+        if (
+            isinstance(key, cudf.core.column.CategoricalColumn)
+            and index.dtype != key.dtype
+        ):
+            # TODO: is this right?
+            key = key._get_decategorized_column()
+        if is_bool_dtype(key.dtype):
+            # The only easy one.
+            return MaskIndexer(BooleanMask(key, n))
+        elif len(key) == 0:
+            return EmptyIndexer()
+        else:
+            # TODO: promote to Index objects, so this can handle
+            # categoricals correctly?
+            (haystack,) = index._columns
+            if isinstance(index, cudf.DatetimeIndex):
+                # Try to turn strings into datetimes
+                key = cudf.core.column.as_column(key, dtype=index.dtype)
+            gather_map = ordered_find(key, haystack)
+            if is_scalar and len(gather_map.column) == 1:
+                return ScalarIndexer(gather_map)
+            else:
+                return MapIndexer(gather_map)
+
+
+def parse_row_loc_indexer(key: Any, index: cudf.BaseIndex) -> IndexingSpec:
+    """
+    Normalize to return structured information for a label-based row indexer.
+
+    Given a label-based row indexer that has already been destructured by
+    :func:`destructure_loc_key`, inspect further and produce structured
+    information for indexing operations to act upon.
+
+    Parameters
+    ----------
+    key
+        Suitably destructured key for row indexing
+    index
+        Index to provide context
+
+    Returns
+    -------
+    IndexingSpec
+        Structured data for indexing. A tag + parsed data.
+
+    Raises
+    ------
+    KeyError
+        If a valid type of indexer is provided, but not all keys are
+        found
+    TypeError
+        If the indexing key is otherwise invalid.
+    """
+    if isinstance(index, cudf.MultiIndex):
+        raise NotImplementedError(
+            "This code path is not designed for multiindices"
+        )
```

Comment:
There are two things that need to happen in this case:
1. individual keys can have more cases (e.g. `(a, b):(c, d)` is a valid slice for a multiindex if the multiindex is ordered), a list-like can be not just single-level labels, but tuples of labels)
2. If we get a tuple of per-level indexers we must combine them to a single output indexer and the semantics of this combination in pandas are ambiguous in many cases.